lib/services/namespace_service.ts

```diff
@@ -0,0 +1,131 @@
+import { DeviceInterface } from '@Lib/device_interface';
+import { Uuid } from '@Lib/uuid';
+import { UuidString } from './../types';
+
+export type SNNamespace = {
+  identifier: string | UuidString,
+  userUuid?: UuidString,
+  label: string,
+  isDefault: boolean,
+};
+
+const NAMESPACES_STORAGE_KEY = 'namespaces';
+
+/**
+ * The namespace service is responsible of setting the namespace to be used by
+ * the DeviceInterface.
+ */
+export class SNNamespaceService extends PureService {
+  private namespace?: SNNamespace
+
+  constructor(deviceInterface: DeviceInterface, namespaceIdentifier?: string) {
+    super();
+    this.deviceInterface = deviceInterface;
+
+    if (namespaceIdentifier) {
+      const namespace: SNNamespace = {
+        identifier: namespaceIdentifier,
+        label: namespaceIdentifier,
+        isDefault: true
+      };
+      this.namespace = namespace;
+      this.deviceInterface!.switchToNamespace(namespace);
+    }
+  }
+
+  private async getNamespaces() {
+    const namespaces = await this.deviceInterface!.getRawStorageValue(NAMESPACES_STORAGE_KEY);
+    if (!namespaces) {
+      return [] as SNNamespace[];
+    }
+    return JSON.parse(namespaces) as SNNamespace[];
+  }
+
+  private async setNamespaces(namespaces: SNNamespace[]) {
+    this.deviceInterface!.setRawStorageValue(
+      NAMESPACES_STORAGE_KEY,
+      JSON.stringify(namespaces)
+    );
+  }
+
+  private async getDefaultNamespace() {
+    const namespaces = await this.getNamespaces();
+    return namespaces.find(namespace => namespace.isDefault === true);
+  }
+
+  private async pushNamespace(namespace: SNNamespace) {
+    const namespaces = await this.getNamespaces();
+    const namespaceIndex = namespaces.findIndex(n => n.identifier === namespace.identifier);
+    if (namespaceIndex === -1) {
+      namespaces.push(namespace);
+    } else {
+      namespaces[namespaceIndex] = namespace;
+    }
+    await this.setNamespaces(namespaces);
+  }
+
+  private async switchToNamespace(namespace: SNNamespace) {
+    this.namespace = namespace;
+    this.deviceInterface!.switchToNamespace(namespace);
+    await this.pushNamespace(namespace);
+  }
+
+  /**
+   * Creates a new namespace if necessary. If not, use an existing one.
+   */
+  public async initialize() {
+    const defaultNamespace = await this.getDefaultNamespace();
+    if (defaultNamespace) {
+      await this.switchToNamespace(defaultNamespace);
+      return;
+    }
+    const namespaces = await this.getNamespaces();
+    /** If no namespaces exist, then we should create a new one. */
+    if (namespaces.length === 0) {
+      await this.createNamespace(true);
+    }
+    /**
+     * If one (1) namespace exist, then it means it is not the default one at this point.
+     * Let's set it as the default namespace and return it.
+     */
+    else if (namespaces.length === 1) {
+      const namespace = namespaces[0];
+      namespace.isDefault = true;
+      namespace.label = 'Default namespace';
```

Comment:
"namespace" is an internal concept. For UI purposes, they should be called "applications". So this would be called "Default application"